File: menu.go
package gadget

import (
	"fmt"

	"github.com/techthos/gadget/internal/htmlx"
	"github.com/techthos/gadget/theme"
	"github.com/techthos/gadget/uispec"
)

// MenuItem is one entry in a Menu: a tool the user can start from the menu
// tile grid. Choosing it calls Tool with Args; the host answers by opening
// that tool's own widget, so a menu item is navigation rather than an action
// with a result of its own.
//
// That relies on the host opening the widget bound to a view-initiated
// tools/call. A host that instead runs such a call out of band answers it
// without opening anything, and the tile looks inert. Prompt switches the item
// to the chat path for those hosts: the host posts the text as a user turn,
// the model calls the tool, and the widget arrives as that call's result.
type MenuItem struct {
	// Tool is the MCP tool called when the item is chosen (required). With
	// Prompt set the call is the model's to make, so Tool documents what the
	// item opens and still supplies the default Label.
	Tool string
	// Args are static arguments passed to Tool. Menu items carry no record,
	// so unlike Action args these are fixed values, not row lookups. Ignored
	// when Prompt is set — the model chooses the arguments there.
	Args map[string]any
	// Prompt, when set, makes the item ask the host to post this text as a
	// user message (ui/message) instead of calling Tool itself. Write it as
	// the request a user would type: the model reads it and decides which
	// tool answers, so name what the item opens rather than the tool.
	Prompt string
	// Label is the tile heading. Defaults to Tool.
	Label string
	// Description is the supporting line under the label.
	Description string
	// IconSVG is inline <svg> markup shown above the label. Documents stay
	// self-contained, so an icon is never a URL. It is author-trusted input,
	// checked only against script-bearing and resource-loading constructs.
	IconSVG string
	// Badge is optional short text shown in the tile's top right — a category
	// or state marker such as "read" or "beta".
	Badge string
	// BadgeVariant colors the badge. Defaults to BadgeNeutral.
	BadgeVariant BadgeVariant
}

func (i MenuItem) label() string {
	if i.Label != "" {
		return i.Label
	}
	return i.Tool
}

// badgeVariants are the values BadgeVariant may take. A menu badge's variant
// becomes a class name in server-rendered markup, so it is checked up front
// rather than silently producing a class no stylesheet rule matches.
var badgeVariants = map[BadgeVariant]bool{
	BadgeNeutral: true,
	BadgeInfo:    true,
	BadgeSuccess: true,
	BadgeWarning: true,
	BadgeDanger:  true,
}

func (i MenuItem) validate(ctx string) error {
	if i.Tool == "" {
		return fmt.Errorf("%s: Tool is required", ctx)
	}
	if i.IconSVG != "" {
		if _, err := htmlx.RawSVG(i.IconSVG); err != nil {
			return fmt.Errorf("%s: IconSVG: %w", ctx, err)
		}
	}
	if i.BadgeVariant != "" && !badgeVariants[i.BadgeVariant] {
		return fmt.Errorf("%s: unknown badge variant %q", ctx, i.BadgeVariant)
	}
	return nil
}

// Menu renders a launcher: a grid of tiles, one per MCP tool the application
// exposes with a UI. Choosing a tile calls that tool, and the host opens the
// widget bound to it — so a Menu is the entry point an app hands the user
// before any particular record is in view.
//
// Unlike the data widgets, a Menu is fully authored at registration time: the
// tiles are server-rendered from Items and the document needs nothing from
// the runtime data contract.
type Menu struct {
	// URI is the widget's ui:// resource URI (required).
	URI string
	// Title is shown in the toolbar and the document title.
	Title string
	// Intro is optional lead text shown above the tiles.
	Intro string
	// Items are the menu entries (at least one is required).
	Items []MenuItem

	// Brand renders the application logo/name on the widget.
	Brand *Brand
	// Theme overrides gadget design tokens for this widget.
	Theme *theme.Theme
	// UI overrides resource _meta.ui (CSP, permissions, prefersBorder).
	UI *uispec.ResourceUIMeta
}

// Validate implements Widget.
func (m *Menu) Validate() error {
	if err := uispec.ValidateURI(m.URI); err != nil {
		return fmt.Errorf("gadget: menu: %w", err)
	}
	if len(m.Items) == 0 {
		return fmt.Errorf("gadget: menu %s: at least one item is required", m.URI)
	}
	for i, item := range m.Items {
		ctx := fmt.Sprintf("gadget: menu %s: item %d (%s)", m.URI, i, item.label())
		if err := item.validate(ctx); err != nil {
			return err
		}
	}
	if err := m.Brand.Validate(); err != nil {
		return fmt.Errorf("gadget: menu %s: %w", m.URI, err)
	}
	if err := m.Theme.Validate(); err != nil {
		return fmt.Errorf("gadget: menu %s: %w", m.URI, err)
	}
	return nil
}

// Descriptor implements Widget.
func (m *Menu) Descriptor() uispec.ResourceDescriptor {
	return uispec.ResourceDescriptor{
		URI:      m.URI,
		Name:     resourceName(m.URI),
		Title:    m.Title,
		MIMEType: uispec.MIMEType,
		UI:       m.UI,
	}
}

// ToolMeta implements Widget.
func (m *Menu) ToolMeta() map[string]any {
	return uispec.ToolUIMeta{ResourceURI: m.URI}.MetaMap()
}

// config serializes only what the runtime needs: the tool call behind each
// tile, positionally matched to the server-rendered buttons. Labels, icons
// and descriptions are already in the markup.
func (m *Menu) config() map[string]any {
	items := make([]map[string]any, len(m.Items))
	for i, item := range m.Items {
		entry := map[string]any{"tool": item.Tool}
		if item.Prompt != "" {
			// A prompt item never calls the tool itself, so its args would be
			// dead weight in the island.
			entry["prompt"] = item.Prompt
		} else if len(item.Args) > 0 {
			entry["args"] = item.Args
		}
		items[i] = entry
	}
	return map[string]any{
		"widget": "menu",
		"items":  items,
	}
}

File: ui/src/widgets/menu.ts
// Menu widget behavior: a launcher grid. Every tile is already in the markup
// (Go renders it from the authored items), so this only maps a click to the
// tool call recorded at the same index in the config island.
//
// A menu item is navigation: the host answers the call by opening the tool's
// own widget, replacing or stacking this view. So nothing is rendered from
// the result — only progress while the call is in flight, and the failure if
// it comes back as one.
//
// An item carrying a prompt navigates through the host's chat instead: the
// host posts the prompt as a user turn (ui/message) and the model makes the
// call, which is the only path that opens anything on hosts that run a
// view-initiated tools/call out of band.
import type { MountContext } from "../index";
import { delegate } from "../dom";
import { M } from "../protocol";
import { errorText, textOf } from "../status";

interface MenuItemCfg {
	tool: string;
	args?: Record<string, unknown>;
	prompt?: string;
}

interface MenuCfg {
	widget: string;
	items: MenuItemCfg[];
}

export function mountMenu(ctx: MountContext): void {
	const cfg = ctx.config as unknown as MenuCfg;
	const { root, bridge } = ctx;

	const items = Array.isArray(cfg.items) ? cfg.items : [];
	const tiles = [...root.querySelectorAll<HTMLButtonElement>("[data-gomu-menu-item]")];
	const statusEl = root.querySelector<HTMLElement>("[data-gomu-status]");

	let busy = false;

	function showStatus(kind: "loading" | "error" | "", msg: string): void {
		if (!statusEl) return;
		statusEl.hidden = msg === "";
		statusEl.textContent = msg;
		statusEl.className = "gomu-status" + (kind ? ` gomu-status--${kind}` : "");
	}

	// The whole menu goes inert during a call: a second tile would race the
	// first one's view swap.
	function setBusy(value: boolean): void {
		busy = value;
		for (const tile of tiles) tile.disabled = value;
	}

	function labelOf(tile: HTMLElement): string {
		return tile.querySelector(".gomu-menu-label")?.textContent ?? "this";
	}

	async function open(item: MenuItemCfg, label: string): Promise<void> {
		if (busy || !item.tool) return;
		setBusy(true);
		showStatus("loading", `Opening ${label}…`);
		try {
			// ui/message carries no tool result: the request resolving means the
			// host accepted the turn, and what opens is the model's answer to it.
			if (item.prompt) {
				await bridge.sendMessage(item.prompt);
				setBusy(false);
				showStatus("", "");
				return;
			}
			const res = await bridge.callTool(item.tool, item.args ?? {});
			setBusy(false);
			if (res.isError) {
				showStatus("error", textOf(res) ?? `Could not open ${label}.`);
			} else {
				showStatus("", "");
			}
		} catch (e) {
			setBusy(false);
			showStatus("error", errorText(e, `Could not open ${label}.`));
		}
	}

	delegate(root, "click", "menu-item", (el, value) => {
		const item = items[Number(value)];
		if (item) void open(item, labelOf(el));
	});

	// A host that keeps this view alive still pushes the tool lifecycle
	// notifications; either one means the call is no longer ours to wait on.
	bridge.on(M.toolResult, () => {
		setBusy(false);
		showStatus("", "");
	});
	bridge.on(M.toolCancelled, () => {
		setBusy(false);
		showStatus("", "");
	});
}
